src/modules/hip/kernel/gaussian_filter.hpp

```diff
@@ -0,0 +1,2235 @@
+                                               uint padLength,
+                                               uint2 tileSize,
+                                               RpptROIPtr roiTensorPtrSrc,
+                                               d_float49 *filterTensor)
+{
+    int hipThreadIdx_x8 = hipThreadIdx_x << 3;
+    int id_x_o = (hipBlockIdx_x * tileSize.x * 8) + hipThreadIdx_x8;
+    int id_y_o = hipBlockIdx_y * tileSize.y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+
+    int id_x_i = id_x_o - padLength;
+    int id_y_i = id_y_o - padLength;
+    d_float24 sum_f24;
+    __shared__ uchar srcShared[SHARED_MEMORY_LENGTH_Y_3C][SHARED_MEMORY_LENGTH_X];
+
+    int srcIdx = (id_z * srcStridesNH.x) + ((id_y_i + roiTensorPtrSrc[id_z].xywhROI.xy.y) * srcStridesNH.y) + ((id_x_i + roiTensorPtrSrc[id_z].xywhROI.xy.x) * 3);
+    int dstIdx = (id_z * dstStridesNH.x) + (id_y_o * dstStridesNH.y) + id_x_o * 3;
+    d_float49 filter_f49 = filterTensor[id_z];
+    float *filter_row1 = &filter_f49.f1[0];
+    float *filter_row2 = &filter_f49.f1[7];
+    float *filter_row3 = &filter_f49.f1[14];
+    float *filter_row4 = &filter_f49.f1[21];
+    float *filter_row5 = &filter_f49.f1[28];
+    float *filter_row6 = &filter_f49.f1[35];
+    float *filter_row7 = &filter_f49.f1[42];
+    sum_f24.f4[0] = static_cast<float4>(0);
+    sum_f24.f4[1] = static_cast<float4>(0);
+    sum_f24.f4[2] = static_cast<float4>(0);
+    sum_f24.f4[3] = static_cast<float4>(0);
+    sum_f24.f4[4] = static_cast<float4>(0);
+    sum_f24.f4[5] = static_cast<float4>(0);
+
+    int3 hipThreadIdx_y_channel;
+    hipThreadIdx_y_channel.x = hipThreadIdx_y;
+    hipThreadIdx_y_channel.y = hipThreadIdx_y + 16;
+    hipThreadIdx_y_channel.z = hipThreadIdx_y + 32;
+
+    uchar *srcShared_channel[3];
+    srcShared_channel[0] = &srcShared[hipThreadIdx_y_channel.x][hipThreadIdx_x8];
+    srcShared_channel[1] = &srcShared[hipThreadIdx_y_channel.y][hipThreadIdx_x8];
+    srcShared_channel[2] = &srcShared[hipThreadIdx_y_channel.z][hipThreadIdx_x8];
+
+    if ((id_x_i >= -(int)padLength) && (id_x_i < roiTensorPtrSrc[id_z].xywhROI.roiWidth) &&
+        (id_y_i >= 0) && (id_y_i < roiTensorPtrSrc[id_z].xywhROI.roiHeight))
+    {
+        rpp_hip_load24_pkd3_to_uchar8_pln3(srcPtr + srcIdx, srcShared_channel);
+    }
+    else
+    {
+        *(reinterpret_cast<uint2 *>(srcShared_channel[0])) = static_cast<uint2>(0);
+        *(reinterpret_cast<uint2 *>(srcShared_channel[1])) = static_cast<uint2>(0);
+        *(reinterpret_cast<uint2 *>(srcShared_channel[2])) = static_cast<uint2>(0);
+    }
+    __syncthreads();
+    if ((id_x_o < roiTensorPtrSrc[id_z].xywhROI.roiWidth) &&
+        (id_y_o < roiTensorPtrSrc[id_z].xywhROI.roiHeight) &&
+        (hipThreadIdx_x < tileSize.x) &&
+        (hipThreadIdx_y < tileSize.y))
+    {
+        gaussian_filter_7x7_row_hip_compute(&srcShared[hipThreadIdx_y_channel.x    ][hipThreadIdx_x8], &sum_f24.f8[0], filter_row1);
```

Comment:
@asalmanp : Can you please review this code. I am not sure why this is called many times and not done in parallel

Reply:
@rrawther for this one, I have conceptually not changed the baseline box_filter kernel from last year, and gaussian_filter uses the same:
[rpp/src/modules/hip/kernel/box_filter.hpp at master · GPUOpen-ProfessionalCompute-Libraries/rpp (github.com)](https://github.com/GPUOpen-ProfessionalCompute-Libraries/rpp/blob/master/src/modules/hip/kernel/box_filter.hpp#L480)

It's actually doing the compute in parallel in different thread blocks. Just that in this case for 7x7 we have 7 rows to process, and each row has 3 colors R/G/B. So the common code helper function needs to be called 7x3=21 times